contracts/Remittance.sol

```diff
@@ -1,16 +1,137 @@
+
+        require(remittanceInstance.funder == address(0), "Remittance already exists");
+
+        remittanceInstance.funder = msg.sender;
+        remittanceInstance.fundsOwed = msg.value;
+        remittanceInstance.expiryDate = expiryDate;
+
+        emit RemittanceFundsCreated(recipientPasswordHash, msg.sender, broker, msg.value, expiryDate);
+
+        return true;
+    }
+
+    // The broker sends the recipient's password to release the funds
+    function release(bytes32 recipientPasswordHash) external whenNotPaused returns(bool success) {
+
+        bytes32 rehash = keccak256PasswordHashBroker(recipientPasswordHash, address(msg.sender));
+
+        //Retrieve remittance
+        RemittanceInstance storage remittanceInstance = remittances[rehash];
+
+        uint256 fundsOwed = remittanceInstance.fundsOwed;
+
+        require(fundsOwed > 0, "No funds available");
+        require(block.timestamp < remittanceInstance.expiryDate, "Remittance has expired");
+
+        // Indicate remittance has been collected
+        remittanceInstance.fundsOwed = 0;
+        remittanceInstance.expiryDate = 0;
+
+        emit RemittanceFundsReleased(rehash, msg.sender, fundsOwed);
+
+        (success, ) = msg.sender.call{value: fundsOwed}("");
+        require(success, "Transfer failed");
+    }
+
+    // Funder can retrieve funds if not claimed by the expiry date
+    function reclaim(bytes32 recipientPasswordHash) external whenNotPaused returns(bool success) {
+
+        checkEmptyHash(recipientPasswordHash);
+
+        //Retrieve remittance
+        RemittanceInstance storage remittanceInstance = remittances[recipientPasswordHash];
+
+        uint256 fundsOwed = remittanceInstance.fundsOwed;
+
+        require(fundsOwed > 0, "No funds available");
+        require(block.timestamp >= remittanceInstance.expiryDate, "The remittance has not expired");
+        require(msg.sender == remittanceInstance.funder, "Only the funder can reclaim funds");
+
+        // Indicate that it was the funder who retrieved the expired funds
+        remittanceInstance.fundsOwed = 0;
+        remittanceInstance.expiryDate = 0;
+
+        emit RemittanceFundsReclaimed(recipientPasswordHash, msg.sender, fundsOwed);
+
+        (success, ) = msg.sender.call{value: fundsOwed}("");
+        require(success, "Transfer failed");
+    }
+
+    // Utility function
+    function keccak256Password(string memory passwordStr) public pure returns(bytes32) {
+        require(bytes(passwordStr).length > 0, 'Non-empty string required');
+
+        return keccak256(abi.encodePacked(passwordStr));
+    }
+
+    // Utility function
+    function keccak256PasswordHashBroker(bytes32 passwordHash, address broker) public pure returns(bytes32) {
+        checkEmptyHash(passwordHash);
+        require(broker != address(0), "Address cannot be zero");
+
+        return keccak256(abi.encodePacked(passwordHash, broker));
```

Comment:
> You don't want passwords to be reused, right? That's why you leave `.funder` in the mapping, so that it cannot be reused.

Yes and to ensure only the funder can reclaim.

> But what happens if the same contract is deployed elsewhere, then the same people can reuse the same passwords. How would you protect against that?

OK. So that means the broker could try to match the remittance hash when contract2 is created with the existing hashes in contract1 and thus know the password is the same?

In which case we can perform a further rehash and use the contract's address like a salt to make it unique for that contract. Would that be a suitable solution?

Reply:
> In which case we can perform a further rehash and use the contract's address like a salt to make it unique for that contract. Would that be a suitable solution?

Yep, sounds like a viable solution.